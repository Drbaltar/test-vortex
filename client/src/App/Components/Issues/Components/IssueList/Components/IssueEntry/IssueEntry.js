import React from 'react';
import ListEntry from '../../../../../shared-components/ListEntry/ListEntry';

const IssueEntry = (props) => {
    const { entry, index, select } = props;

    return <ListEntry entryFields={getIssueFields(entry)} index={index} select={select}/>
};

const getIssueFields = (issue) => {
    return [
        {
            label: 'Issue Type:',
            field: issue.issue_type,
        },
        {
            label: 'Issue Category:',
            field: issue.issue_category
        },
        {
            label: 'Issue Description:',
            field: issue.issue_description
        },
        {
            label: 'Submitted By:',
            field: issue.submitted_by
        }
    ];
}; 

export default IssueEntry;
