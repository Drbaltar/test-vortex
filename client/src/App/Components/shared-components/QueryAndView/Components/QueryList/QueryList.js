import React from 'react';

const QueryList = (props) => {
    const { list, entryType, selectEntry } = props;
    const Entry = props.children;

    return(
        <div> 
            <p id='query-list-message'>{getErrorMessage(entryType, list.length)}</p>
            {getList(Entry, list, selectEntry)}
        </div>
    );
};

const getErrorMessage = (entryType, length) => {
    if (length === 0) {
        return `No ${entryType}s Found!`;
    } else {
        return `${length} ${entryType}${length > 1 ? 's' : ''} Found!`;
    }
};

const getList = (Entry, list, selectEntry) => {
    return list.map((entry, index) => {
        return React.cloneElement(Entry,
            { 
                entry: entry,
                key: index,
                select: selectEntry,
                index
            }
        );
    });
};

QueryList.defaultProps = {
    list: []
};

export default QueryList;
