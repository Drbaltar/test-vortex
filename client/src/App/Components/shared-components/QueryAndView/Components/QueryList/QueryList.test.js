import React from 'react';
import { shallow } from 'enzyme';
import QueryList from './QueryList';
import IssueEntry from './Components/IssueEntry/IssueEntry';

const singleEntry = [
    {
        issue_type: 'Application Issue',
        issue_category: 'Application Crash',
        issue_description: 'The application keeps crashing!',
        submitted_by: 'kmccain'
    }
];

const exampleList = [
    {
        issue_type: 'Application Issue',
        issue_category: 'Application Crash',
        issue_description: 'The application keeps crashing!',
        submitted_by: 'kmccain'
    },
    {
        issue_type: 'Application Issue',
        issue_category: 'Entry/List Not Populating Information',
        issue_description: 'The question list is not populating!',
        submitted_by: 'kmccain'
    }
];

const entryType = 'Question';

const mockSelectFunction = jest.fn();

describe('QueryList', () => {
    const wrapper = shallow(<QueryList entryType={entryType}>
        <IssueEntry />
    </QueryList>);

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe('when an empty list is passed in', () => {
        beforeAll(() => {
            wrapper.setProps({ list: [] });
        });

        it('returns a message that there are no entries found if passed an empty list', () => {
            expect(wrapper.find('#query-list-message').text()).toBe(`No ${entryType}s Found!`);
        });
    });

    describe('when a single item list is passed in', () => {
        beforeAll(() => {
            wrapper.setProps({
                list: singleEntry,
                selectEntry: mockSelectFunction
            });
        });

        it('returns a message that there is one issue found', () => {
            expect(wrapper.find('#query-list-message').text()).toBe(`1 ${entryType} Found!`);
        });
    });

    describe('when a list greater than one entry is passed in', () => {
        beforeAll(() => {
            wrapper.setProps({
                list: exampleList,
                selectEntry: mockSelectFunction
            });
        });

        it('returns a message that there are `n` number of entries found', () => {
            expect(wrapper.find('#query-list-message').text()).toBe(`${exampleList.length} ${entryType}s Found!`);
        });

        it('creates the correct number of list items', () => {
            expect(wrapper.find('IssueEntry').length).toEqual(exampleList.length);
        });

        it('passes the issue entry to the IssueEntry component', () => {
            expect(wrapper.find('IssueEntry').first().prop('entry')).toEqual(exampleList[0]);
        });

        it('passes the index entry to the IssueEntry component', () => {
            expect(wrapper.find('IssueEntry').at(0).prop('index')).toEqual(0);
            expect(wrapper.find('IssueEntry').at(1).prop('index')).toEqual(1);
        });

        it('passes the function for selecting an entry to the IssueEntry component', () => {
            expect(wrapper.find('IssueEntry').first().prop('select')).toEqual(mockSelectFunction);
        });
    });

});