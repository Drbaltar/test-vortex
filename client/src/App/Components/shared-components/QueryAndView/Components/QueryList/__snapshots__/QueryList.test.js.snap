// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryList renders correctly 1`] = `
<div>
  <p
    id="query-list-message"
  >
    No Questions Found!
  </p>
</div>
`;
